cartridge/pool.lua

```diff
@@ -193,25 +197,58 @@ local function map_call(fn_name, args, opts)
     end
 
     local retmap, errmap = {}, {}
-    local fibers = {}
-    for _, uri in pairs(opts.uri_list) do
-        local fiber = fiber.new(
-            _gather_netbox_call,
-            fiber.self().storage, retmap, errmap, uri, fn_name, args,
-            {timeout = opts.timeout}
-        )
-        fiber:name('netbox_map_call')
-        fiber:set_joinable(true)
-        fibers[uri] = fiber
+    local fibers = table.new(0, #opts.uri_list)
+    local futures = table.new(0, #opts.uri_list)
+
+    local timeout = opts.timeout or vars.options.MAP_CALL_TIMEOUT
+    local deadline = fiber.clock() + timeout
+
+    for _, uri in ipairs(opts.uri_list) do
+        local conn, err = connect(uri, {wait_connected = false})
+        if conn == nil then
+            err = err or 'Unknown error'
+            errmap[uri] = NetboxConnectError:new('%q: %s', uri, err)
+        elseif conn:is_connected() then
+            local future, err = errors.netbox_call(
+                conn, fn_name, args, {is_async = true}
+            )
+            if future == nil then
+                errmap[uri] = err
+            end
+            futures[uri] = future
+        else
+            local fiber = fiber.new(_gather_netbox_call,
+                fiber.self().storage,
+                retmap, errmap,
+                conn, uri,
+                fn_name, args,
+                deadline
+            )
+            fiber:name('netbox_map_call')
+            fiber:set_joinable(true)
+            fibers[uri] = fiber
+        end
     end
 
-    for _, uri in pairs(opts.uri_list) do
-        local ok, err = fibers[uri]:join()
+    for uri, fiber in pairs(fibers) do
+        local ok, err = fiber:join()
         if not ok then
             errmap[uri] = NetboxMapCallError:new(err)
         end
     end
 
+    for uri, future in pairs(futures) do
+        local timeout = deadline - fiber.clock()
+        if timeout < 0 then
+            timeout = 0
+        end
+
+        local ret, err = errors.netbox_wait_async(future, timeout)
+        retmap[uri] = ret
+        errmap[uri] = err
+        future:discard()
```

Comment:
IFIU discard is needed for "in-fly" requests. In case of successful (or finished with error) result it doesn't make much sense.

Reply:
Yes, it's a case when the `netbox_wait_async` times out.